SDL/DescriptorDb.cpp

```diff
@@ -88,39 +90,154 @@ SDL::StateDescriptor* SDL::DescriptorDb::FindDescriptor(const ST::string& name,
+        case SDL::e_VarVector3:
+            return ST_LITERAL("VECTOR3");
+        case SDL::e_VarPoint3:
+            return ST_LITERAL("POINT3");
+        case SDL::e_VarRgb:
+            return ST_LITERAL("RGB");
+        case SDL::e_VarRgba:
+            return ST_LITERAL("RGBA");
+        case SDL::e_VarQuaternion:
+            return ST_LITERAL("QUATERNION");
+        case SDL::e_VarRgb8:
+            return ST_LITERAL("RGB8");
+        case SDL::e_VarRgba8:
+            return ST_LITERAL("RGBA8");
+        default:
+            return ST_LITERAL("INVALID");
+    }
+}
+
+static inline ST::string default_to_str(const SDL::VarDescriptor& var)
+{
+    if (!var.m_default.m_valid)
+        return {};
+
+    switch (var.m_type) {
+        case SDL::e_VarInt:
+        case SDL::e_VarBool:
+        case SDL::e_VarByte:
+        case SDL::e_VarShort:
+            return ST::string::from_int(var.m_default.m_int);
+        case SDL::e_VarFloat:
+            return ST::string::from_float(var.m_default.m_float);
+        case SDL::e_VarString:
+            return var.m_default.m_string;
+        case SDL::e_VarKey:
+            // The only valid default is "nil", which is the same as leaving the default blank.
+            // So, let's just save these bytes.
+            return {};
+        case SDL::e_VarDouble:
+            return ST::string::from_double(var.m_default.m_double);
+        case SDL::e_VarTime:
+            return ST::string::from_uint(var.m_default.m_time.m_secs);
+        case SDL::e_VarVector3:
+        case SDL::e_VarPoint3:
+            return ST::format("({},{},{})", var.m_default.m_vector.m_X,
+                              var.m_default.m_vector.m_Y, var.m_default.m_vector.m_Z);
+        case SDL::e_VarRgb:
+            return ST::format("({},{},{})", var.m_default.m_color.m_R,
+                              var.m_default.m_color.m_G, var.m_default.m_color.m_B);
+        case SDL::e_VarRgba:
+            return ST::format("({},{},{},{})", var.m_default.m_color.m_R,
+                              var.m_default.m_color.m_G, var.m_default.m_color.m_B,
+                              var.m_default.m_color.m_A);
+        case SDL::e_VarRgb8:
+            return ST::format("({},{},{})", var.m_default.m_color8.m_R,
+                              var.m_default.m_color8.m_G, var.m_default.m_color8.m_B);
+        case SDL::e_VarRgba8:
+            return ST::format("({},{},{},{})", var.m_default.m_color8.m_R,
+                              var.m_default.m_color8.m_G, var.m_default.m_color8.m_B,
+                              var.m_default.m_color8.m_A);
+        case SDL::e_VarQuaternion:
+            return ST::format("({},{},{},{})", var.m_default.m_quat.m_X,
+                              var.m_default.m_quat.m_Y, var.m_default.m_quat.m_Z,
+                              var.m_default.m_quat.m_W);
+        default:
+            DS_ASSERT(0);
+    }
+}
+
+void SDL::DescriptorDb::WriteDescriptors(DS::Stream* stream)
```

Comment:
Yikes!  Why not just encrypt the source file, instead of adding all this additional complexity by writing out processed files?  More specifically, since you've added the capability to handle encrypted files, why not just use the encrypted files as a source rather than continuing to store decrypted SDL separately (hint: that's also what I did for moulars ;) )...

Reply:
I'm having trouble parsing the moulars source code atm, so I'm not really sure what's going on over there. I do agree, though, that this method is fairly complex. My goal is to ensure that the SDL that the AuthSrv sends to the client is functionally the same as what is read into the descriptor database. My primary goal with this and the client change is to enforce a contract about what the SDL is - the processing is just a side effect with some benefits and drawbacks (namely, the added complexity). My thinking was that if we don't re-encode the SDL that's in the database but rather simply read it off the disk each time, then we would have a problem with that contract. DS currently initializes the descriptor database at start up and doesn't bother listening for any changes to the SDL directory. If we just send the contents of the SDL directory to client, any changes to the SDL directory would get sent down to the client but be unknown to DS. That really defeats the purpose of this PR, IMO.

Big picture: My eventual goal is to, in future changes, allow DS to listen for changes to its SDL directory. Ideally, this will allow changes to state descriptor versions themselves so that Ages undergoing active development don't require a vault wipe or excessive SDL version increments for minor development churn (this has been a problem for some Age/content creators in the past). To effect this change, DS needs to know the same state descriptors that the client knows so that the two can communicate in the same blob format. Right now, DS isn't able to cope with two version 1s of the same state descriptor, so I thought that round-tripping the SDL would be for the best. There is already precedent for this - DS round-trips everything else, even Plasma messages that could just be broadcast without being checked.

> why not just use the encrypted files as a source

I did update the code to allow the SDL files to be encrypted, but it doesn't seem to me like that's the best default way to store things. Ideally, we just want the SDL to be plaintext on the server for ease of administration.